src/system_tests/component_metadata.cpp

```diff
@@ -0,0 +1,220 @@
+            },
+            "10493596": {
+              "message": "Airspeed sensor failure detected. Return to launch (RTL) is advised",
+              "name": "airspeed_selector_sensor_failure"
+            }
+        }
+      }
+    }
+  },
+  "version": 2
+}
+)JSON";
+
+TEST(SystemTest, ComponentInformationConnect)
+{
+    Mavsdk mavsdk_groundstation{Mavsdk::Configuration{Mavsdk::ComponentType::GroundStation}};
+    ASSERT_EQ(mavsdk_groundstation.add_any_connection("udp://:17000"), ConnectionResult::Success);
+
+    Mavsdk mavsdk_companion{Mavsdk::Configuration{Mavsdk::ComponentType::CompanionComputer}};
+    ASSERT_EQ(
+        mavsdk_companion.add_any_connection("udp://127.0.0.1:17000"), ConnectionResult::Success);
+
+    // Wait for companion system to connect
+    auto prom = std::promise<std::shared_ptr<System>>();
+    std::once_flag flag;
+    mavsdk_groundstation.subscribe_on_new_system([&]() {
+        std::call_once(flag, [&prom, &mavsdk_groundstation]() {
+            prom.set_value(mavsdk_groundstation.systems().at(0));
+        });
+    });
+    auto fut = prom.get_future();
+    ASSERT_EQ(fut.wait_for(std::chrono::seconds(5)), std::future_status::ready);
+    auto system = fut.get();
+
+    auto server = ComponentMetadataServer{mavsdk_companion.server_component()};
+
+    // Register metadata
+    std::vector<ComponentMetadataServer::Metadata> metadata;
+    metadata.push_back(
+        {ComponentMetadataServer::MetadataType::Parameter, std::string(parameter_json_metadata)});
+    metadata.push_back(
+        {ComponentMetadataServer::MetadataType::Events, std::string(events_json_metadata)});
+    server.set_metadata(metadata);
+
+    // Ask for metadata
+    auto client = ComponentMetadata{system, {MAV_COMP_ID_ONBOARD_COMPUTER}};
+    std::atomic_bool received_parameters{false};
+    std::atomic_bool received_events{false};
+    client.register_notification_callback(
+        [&received_events, &received_parameters](
+            ComponentMetadata::MetadataType type, uint32_t compid, const std::string& json_data) {
+            LogInfo() << "Got metadata, type: " << static_cast<int>(type);
+            EXPECT_EQ(compid, MAV_COMP_ID_ONBOARD_COMPUTER);
+            switch (type) {
+                case ComponentMetadata::MetadataType::Parameter:
+                    EXPECT_EQ(json_data, parameter_json_metadata);
+                    received_parameters = true;
+                    break;
+                case ComponentMetadata::MetadataType::Events:
+                    EXPECT_EQ(json_data, events_json_metadata);
+                    received_events = true;
+                    break;
+                default:
+                    ASSERT_TRUE(false) << "Unexpected metadata type " << static_cast<int>(type);
+            }
+        });
+
+    for (int i = 0; i < 100 && (!received_events || !received_parameters); ++i) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
```

Comment:
You didn't want to do this one with `std::future` and `std::promise`?

Reply:
I thought it's not worth the extra complexity given it's a test and we need to wait for multiple things.

Reply:
Right, I wonder what the canonical way is to wait for two futures.